### DISCUSSION TITLE
Determine if a CadDocument contains proxy objects or entities

### DISCUSSION THREAD

**Discussion by thorster** [DomCR/ACadSharp#271] [Q&A]
How can I determine if a CadDocument contains proxies? Tried this but does not seem to work:

```
foreach (Entity ent in cadDoc.Entities)
{
    if (ent.ObjectType is ObjectType.ACAD_PROXY_ENTITY or ObjectType.ACAD_PROXY_OBJECT)
    {
        Debug.WriteLine(ent.ObjectName); 
    }
}
```

**Comment by DomCR** [comment]
Hi @thorster 

Proxies are not currently supported by `ACadSharp`, I can open a branch to implement them in the future, but the documentation on proxies is quite limited, I'll have to check if `ACadSharp` can offer a full support on them.

**Reply by DomCR** [reply]
- https://github.com/DomCR/ACadSharp/issues/272

**Reply by thorster** [reply]
Thanks, at this point I only want to detect if there are any proxies when the dwg is read. For this purpose it is not necessary to read and create the objects in the CadDocument. I have tried this:

![image](https://github.com/DomCR/ACadSharp/assets/4280346/108378bd-655e-4074-9c67-a173aba04234)

But this will not cover all proxies as many proxies in my drawings does not have types `ObjectType.ACAD_PROXY_ENTITY` or `ObjectType.ACAD_PROXY_OBJECT`  instead they have "unlisted" types. I have .NET plugin code that lists proxies, This is the result for the uploaded dwg-file:
 

> Total Proxies: 108
>       Proxy Entities: 25
>         "ImpactRcBase|Application: IMPACT Reinforcement|Copyright Structural Design Software in Europe AB 2003|http://www.strusoft.com": 24
>           PointAecRcScheduleEntity (IMPACTRCSCHEDULEENTITY): 1
>           PointAecRcRebar (REINFORCEMENTBAR): 23
>         "PrecastBase" "|Product Desc:     IMPACT Precast" "|Company:          StruSoft" "|WEB Address:      www.strusoft.com": 1
>           ImpPrecastRcAreaMesh (ImpPrecastRcAreaMesh): 1
>       Proxy Objects: 83
>         "ImpactRcBase|Application: IMPACT Reinforcement|Copyright Structural Design Software in Europe AB 2003|http://www.strusoft.com": 83
>           PointAecRcSchedule (IMPACTRCSCHEDULE): 1
>           PointAecRcTemplateData (IMPACTRCTEMPLATEDATA): 55
>           PointAecRcBarCode (IMPACTRCBARCODE): 17
>           PointAecRcBarDef (IMPACTRCBARDEF): 5
>           PointAecRcQuality (IMPACTRCQUALITY): 1
>           PointAecRcStyle (IMPACTRCSTYLE): 2
>           PointAecRcBarEnd (IMPACTRCBAREND): 1
>           PointAecRcDatabase (IMPACTRCDATABASE): 1

When this dwg-file [proxies.zip](https://github.com/DomCR/ACadSharp/files/14177814/proxies.zip) is read with `ACadSharp` the proxies are not detected as proxies, instead they have unlisted types, Is it possible without to much effort to detect as proxies?

**Reply by DomCR** [reply]
You could use the notification system to detect the proxies, if you detect a notification that contains the string "PROXY" in the message then it means that the file has proxies in it.

**Reply by thorster** [reply]
I have tried that but as mentioned `ACadSharp` does not report that the proxies in the sample file actually are proxies. Instead it reports them as unlisted types and objects:

> Unlisted type: IMPACTRCTEMPLATEDATA - 518
> Unlisted object not implemented, DXF name: IMPACTRCTEMPLATEDATA`

There are other types and objects that not are proxies so detecting "Unlisted" in the message will include other types and objects.

**Reply by DomCR** [reply]
I've been trying to find a way to detect Proxies in a file but right now is quite hard to detect them, the documentation only gives the class `ACAD_PROXY_ENTITY` type to find them but it seems that there are many others.

Quote from the docs:
> Proxy entities (zombies prior to R14) can have associated graphics data.

Which means that by now a work around could be to check the DxfClasses and look for those that have the `WasZombie` flag set to true, if I check those types manually in a file I can see that they are proxies, also the DxfClasses that are generated by default and linked to a specific Dxf documented object have the flag set to false.

Let me know if this helps, I'll try to gather more information about this issue.

**Reply by DomCR** [reply]
I've merged this PR with a fix for this issue:
- https://github.com/DomCR/ACadSharp/pull/335

When reading a DWG document you can set the flag `DwgReaderConfiguration.KeepUnknownEntities` set to true, this will add all entities found in the file, those entities that are proxies or not supported by `ACadSharp` will be added as `UnknownEntity` and will hold the DxfClass linked to them.

To check if an `UnknownEntity` is a proxy, check the `WasZombie` flag in it's `DxfClass`, if is true, then is a proxy entity.

**NOTE**: the documentation is not clear about the `WasZombie` flag, I've tested different files having proxy entities and the common variable was this flag, but I cannot guarantee if it would be effective in all the cases.
